cpp/src/arrow/dataset/filter.cc

```diff
@@ -839,6 +942,10 @@ struct TreeEvaluator::Impl {
     return this_->Evaluate(expr, batch_);
   }
 
+  Result<Datum> operator()(const Expression& expr) const {
```

Comment:
Is this overload useful? Shouldn't all calls be matched against the templated overload above?

Reply:
This confusion arises from TreeEvaluator's methods all being named `Evaluate`. I'll rename the non overridden ones `DoEvaluate`, which should clear this up